### DISCUSSION TITLE
How to best go about contextual permissions?

### DISCUSSION THREAD

**Discussion by Epigene** [palkan/action_policy#209] [Q&A]
Say my Users can either create a standalone Event or an Event that is a part of a Series, and the policy is slightly different for each case, namely, standalone Events can be created freely, but an Event in a Series can only be created by someone who can edit a series (has :update permission).

I imagine I can do something like this

```rb
class EventPolicy < ApplicationPolicy
  authorize :series, optional: true
  
  def create?
    true
  end 
  
  def create_for_series?
    allowed_to?(:create?) &&  SeriesPolicy.new(user, series: series).update?    
  end
end

EventPolicy.new(user, series: series).create_for_series?
#=> true 
```

But I'm not happy that the whole policy needs to be polluted with an optional context option. Do plain kwarg(s) work? (I'm not particularly worried about how they mess with controller helpers, prefer explicit calls)

```rb
class EventPolicy < ApplicationPolicy
  def create?
    true
  end 
  
  def create_for_series?(series:)
    allowed_to?(:create?) &&  SeriesPolicy.new(user, series: series).update?    
  end
end

EventPolicy.new(user).create_for_series?(series: series)
#=> true
```

**Comment by palkan** [comment]
> I'm not happy that the whole policy needs to be polluted with an optional context option

My rule of thumb: if something makes you unhappy, there is likely a better way to do that 🙂

Dealing with `create`-like rules is one of the corner-stone problems when designing policies. Partially because popular tools like CanCanCan and Pundit suggest using a `ResourcePolicy#create?` (or `can :create?`) for that, and even passing an non-persistent record as an object (which, IMO, is a design smell: polices shouldn't operate on something _new_).

l prefer the following approach: authorize the _parent context_ access. In your case it could be either relying on a general rule, say, `manage?` for the parent resource, i.e., `authorize! series, to: :manage?` (read as "you can only create an event within a series if you can manage this series") ir `authorize! something, to: :manage?` (what is the parent resource for a standalone event? maybe, it could be some virtual model). A slightly modified version of this could be using a specific rule: `authorize! whatever, to: :create_events?`.

**Reply by Epigene** [reply]
I also think that operating on something new is unnecessary.

> I prefer the following approach: authorize the parent context access

I am, in `SeriesPolicy.new(user, series: series).update?`. I'd really like for `create_for_series?` rule to be in `EventPolicy`, since it **is** a rule for event creation, placing it in a different class would obliterate discoverability.

So let me ask again, is there a way to expose the `series` context argument on the rule method only, and not on the whole policy class? And if yes, can you give an example in the explicit, policy initialization format?

**Reply by palkan** [reply]
Currently, the only way to achieve this is to add an optional context to the policy (`authorize :series, optional: true`) and pass it explicitly during authorization:

```ruby
authorize! Event, to: :create_for_series, context: {series:}
```

(The `optional: true` was added as an alternative to more flexible (good) and less-declarative (not-so-good) args support for rules: https://github.com/palkan/action_policy/issues/89)

**Reply by Epigene** [reply]
Seems your first instinct back in 2019 also was rule arguments
```rb
def rule?(args:)
  # do smth
end
```

I wonder, what argued (no pun intended) against them. I don't particularly care about the negative impacts on the auth helper, devs can roll their own.  
Or was it some difficulty with caching results if there are arguments (which brings us the the role of `#apply`)?

**Reply by palkan** [reply]
Yeah, caching/memoization is a particular problem here.

And adding args doesn't really fit the model subject-object-context, which we heavily rely on. Too much flexibility hurts conventions 🤷🏻‍♂️
